pretyping/inductiveops.ml

```diff
@@ -758,8 +758,11 @@ let type_of_projection_knowing_arg env sigma p c ty =
    syntactic conditions *)
 
 let control_only_guard env sigma c =
+  let c = Evarutil.nf_evar sigma c in
```

Comment:
Why don't we just to_constr here?

Reply:
Because the types do not agree. There is no `iter_full_binder` for kernel terms...

Reply:
So add it.

Reply:
In fact it used to be on kernel terms until the commit which introduced this issue.

Reply:
Indeed this is what I was going to say, the origin of this problem was a try not to duplicate that function; it is clear now that the motivation was ill-founded so we are bound to duplication.

Reply:
There are no `full_binders` functions in the kernel, I am not really willing to introduce just that one function there. Either we are uniform and provide this variant for all functions or we keep it for EConstr. Such functions are not required by the kernel so I'd advocate against kernel code inflation.

Reply:
I don't understand this. Why would such combinators be needed on `econstr` but not `constr`? Kernel code inflation is irrelevant here IMHO, the code is generic, doesn't increase the complexity for bugs, and is low risk if tested properly.

Reply:
I am fine with adding it, but then we need all such functions, not just `iter`.

Reply:
I agree, but that's certainly not an argument for not adding one ;)

Reply:
OK, so it turns out it is harder than I thought. In order to implement the `full_binder` variants you need term lifting. But the latter is only defined in `Vars` so that the variants cannot be defined in the `Constr` module. What do?

Reply:
Define it in vars, or move lifting to constr
Or have a module outside the kernel for stuff on constr that the kernel doesn't use.

Reply:
Guys, this is really starting to be very complex for a simple fix that is 3-line patch...

Reply:
Well, indeed; but on the other hand this is highligthing serious issues with the API; given that the fix seems far from "critical" we can take some time to discuss what'd be the proper way to address this.

Reply:
>OK, so it turns out it is harder than I thought. In order to implement the full_binder variants you need term lifting. But the latter is only defined in Vars so that the variants cannot be defined in the Constr module. What do?

Actually I don't see where fold_with_full_binders needs term lifting. It needs Context but that was moved before Constr so it's not a problem.

Reply:
It needs term lifting to push fixpoint and cofixpoint binders. Look at `Environ.push_rec_types`.

Reply:
The current code does 
~~~ocaml
  | Fix (_,(lna,tl,bl)) ->
      let n' = CArray.fold_left2 (fun c n t -> g (LocalAssum (n, t)) c) n lna tl in
      let fd = Array.map2 (fun t b -> (t,b)) tl bl in
      Array.fold_left (fun acc (t,b) -> f n' (f n acc t) b) acc fd
~~~
so is it broken?

Reply:
Looks like it...

Reply:
Interstingly it seems it has been broken since the very beginning, see d9dc86a6f7194c2e7ea704c95495955ca4f4b08d from November 2000 (!).

Reply:
The annoying part is that IIUC once we have rec-rec fixpoints the lift will go away x_x

Reply:
Well, I think we still have time to see that coming.